### PR TITLE
Elc vs MMM2D fixes, tests and docu

### PR DESCRIPTION
**Pull request by KonradBreitsprecher**
Part of #1559 

 - Testcases that compare elc vs mmm2d for inert, dielectric and constant potential BCs for neutral and noneutral systems. The tests pulls a test charge through a fixed configuration of charges and compares forces and energies between mmm2d and elc. The testcase is pretty slow with -np 4 (~2 x 23s)
- Almost compliant interface for setting dielectric contrast and constant potential for the two.
- Sanity check for elc: nonneutral + nonmetallic BCs not allowed
- Sanity check for mmm2d: verlet lists not allowed
- Extended docu covering nonneutral systems and dielectrics for elc.
- Fixed constant potental energy sign issue in elc similar to #1568. Covered by the testcase.
- Fixed MMM2d const pot MPI bug.

### CODE REVIEW

**Comment by RudolfWeeber** [comment]
Could you please integrate this with the mixed periodicity
electrostatics test branch? Some changes seem to be in both.

File: src/python/espressomd/actors.pyx [hunk 1]
```diff
@@ -10,8 +10,7 @@ cdef class Actor(object):
                        MagnetostaticInteraction=False,
                        MagnetostaticExtension=False,
                        HydrodynamicInteraction=False,
-                       ElectrostaticExtensions=False,
-                       Scafacos=True)
```
Comment (fweik) [inline src/python/espressomd/actors.pyx]: If the Scafacos here no longer needed?

File: src/core/mmm2d.cpp [hunk 2]
```diff
@@ -1010,7 +1013,7 @@ static void setup_PQ(int p, int q, double omega, double fac)
   double layer_top;
   double e, e_di_l, e_di_h;
   double *llclcblk;
-  double *lclimgebot=nullptr, *lclimgetop=nullptr;
+  double *lclimgebot=NULL, *lclimgetop=NULL;
```
Comment (fweik) [inline src/core/mmm2d.cpp]: nullptr

File: src/python/espressomd/cellsystem.pyx [hunk 3]
```diff
@@ -199,7 +199,17 @@ cdef class CellSystem(object):
...
-            raise Exception('node_grid is not settable by the user.')
+            if not np.prod(_node_grid) == n_nodes:
+                raise ValueError("Number of available nodes " + str(n_nodes) + " and imposed node grid " + str(_node_grid) + " do not agree.")
+            else:
+                node_grid[0] = _node_grid[0]
+                node_grid[1] = _node_grid[1]
+                node_grid[2] = _node_grid[2]
+                mpi_err = mpi_bcast_parameter(FIELD_NODEGRID)
+                handle_errors("mpi_bcast_parameter failed")
+                #raise Exception('node_grid is not settable by the user.')
```
Comment (fweik) [inline src/python/espressomd/cellsystem.pyx]: commented out code

File: src/python/espressomd/electrostatic_extensions.pyx [hunk 4]
```diff
@@ -30,6 +30,57 @@ IF ELECTROSTATICS and P3M:
...
+                                  explicitely as constraints, you have to disable the neutralization.
+                                  When using a dielectric contrast or full metallic walls (`delta_mid_top
+                                  != 0` or `delta_mid_bot != 0` or `const_pot_on=1`), `neutralize` is
+                                  overwritten and switched off internally. Note that the special case of
+                                  non-neutral systems with a *non-metallic* dielectric jump (eg.
+                                  `delta_mid_top` or `delta_mid_bot` in `]-1,1[`) is not covered by the
+                                  algorithm and will throw an error.
+        far_cut                 : float, optional
+                                  Cut off radius, use with care, intended for testing purposes. 
+        """
```
Comment (fweik) [inline src/python/espressomd/electrostatic_extensions.pyx:83]: iirc there should be an empty line at the end of the docstring.

File: src/python/espressomd/actors.pyx (hunk 1 above)
Comment (KonradBreitsprecher) [reply src/python/espressomd/actors.pyx]: That one came with @RudolfWeeber.

Comment (fweik) [reply src/python/espressomd/actors.pyx]: Rudolf is looking into it.

**Comment by RudolfWeeber** [comment]
When I build this with scafacos, the coulomb cloud wall test segfaults at exit.

**Comment by RudolfWeeber** [comment]
@KonradBreitsprecher @fweik : 
* fixed the segfault (gpu energies being copied to memory that wasn't there)
* reduced elc gap size (your fix fixed the overconfident tuning and it work swith a cubic box and gap size=1)
* Adjusted the field accuracy in the p2nfft test for it to produce sufficient accuracy
* reactivated Scafacos as a there-can-only-be-one actor
Please review (and merge) the branch pr1638 at my gihtub. Then, this pr can be merged, from my side.

**Comment by KonradBreitsprecher** [comment]
Looks good. Tested again with scafacos and got no issues.

**Comment by fweik** [comment]
So ready to merge?

**Comment by RudolfWeeber** [comment]
On Mon, Dec 04, 2017 at 03:46:24PM +0000, Florian Weik wrote:
> So ready to merge?
Yes.

**Comment by RudolfWeeber** [comment]
I pushed fixes for parallel scafacos electrostatics. 
@fweik, if you are happy with the code and the CI tests pass, I think, this is ready.

**Comment by RudolfWeeber** [comment]
I tested with SCAFACOS and with SCAFACOS_DIPOLES coulomb_cloud_wall _mixed_periodicity dipolar_mdlc_scafacos scafacos_1d2d

**Pull request merged**